WIP: verify doctesting refactor

ev-br opened the pull request:
*Work-in-progress, DO NOT MERGE*

Ref: https://mail.python.org/archives/list/scipy-dev@python.org/thread/XH5RDJ5GZRRBP3UAQGWNWMH6P3KYZAUA/

- Split our modified doctesting into a separate package, and add scipy-specific driver to tools. 
- The doctesting internals are still at https://github.com/ev-br/scpdt, need to be installed manually. Naming and plumbing it through <s>do.py</s>dev.py <s>is still pending</s> seems to work modulo gh-16476.
- Patch up `refguide-check.py` to log the names of the objects it examines, do the same for `doctest_public_modules.py` and compare the logs. Locally they are indeed identical, modulo `stats.kstwobign` and `stats.levy_stable` which were skipped in refguide-check. Apparently, they have been fixed at some point, since new doctets pass for them.
- Run the new-style doctesting with warnings turned to errors. This smokes out that our docstrings do emit several warnings, which need to be checked maybe. Also it needs deciding if we actually want to run doctests with warnings being errors.

rgommers reviewed (commented):
Overall LGTM, thanks @ev-br.

One thing that is still missing is how to pull in the new package (git submodule, test dependency, straight-up vendoring), which I guess you're going to do right at the end.

> Run the new-style doctesting with warnings turned to errors. This smokes out that our docstrings do emit several warnings, which need to be checked maybe. Also it needs deciding if we actually want to run doctests with warnings being errors.

This seems reasonable to me. One advantage is that these type of warnings sometimes show up in the Sphinx build too, and there (a) they are already turned into errors, and (b) are way harder to spot and work on.

ev-br commented:
> One thing that is still missing is how to pull in the new package (git submodule, test dependency, straight-up vendoring), which I guess you're going to do right at the end.

Indeed, it is missing :-). Just ran out of time for today, will do later. Vendoring I'd prefrer to do without, so it's either a submodule (did not manage to figure out how to build a submodule: there is a small project-specific script + the submodule, all is in `scipy/tools`) or a test dependency. I noticed that numpydoc is no longer a submodule, so maybe the latter is reasonable. Am undecided yet.

>>Run the new-style doctesting with warnings turned to errors. This smokes out that our docstrings do emit several warnings, which need to be checked maybe. Also it needs deciding if we actually want to run doctests with warnings being errors.

> This seems reasonable to me. One advantage is that these type of warnings sometimes show up in the Sphinx build too, and there (a) they are already turned into errors, and (b) are way harder to spot and work on.

The current testing log (on a released scipy, so might be slightly different) is visible at https://github.com/ev-br/scpdt/runs/6846072299?check_suite_focus=true

ev-br commented on tools/doctest_public_modules.py:
```diff
@@ -0,0 +1,100 @@
+    'constants',     ## obj.__name__
+    'fft',
+    'fftpack',
+    'fftpack.convolve',
```
Thanks for spotting it! The listing further down has a comment 
```
# Docs for these modules are included in the parent module
OTHER_MODULE_DOCS = {
    'fftpack.convolve': 'fftpack',"
```
but the relevant logic from refguide-check is indeed missing. One more thing to fix then :-). Whether to remove it from this list or not is a bit a question: we probably want to keep checking docstrings from this module, or do we not.

rgommers replied:
No need to check separately I'd say. Should be checked as part of `fftpack`, which itself is already legacy. And I don't think I've ever seen `fftpack.convolve` submodule usage in the wild.

ev-br replied:
Hm, the contents of `fftpack.convolve` does not seem to be included into the `fftpack` namespace:

```
In [2]: from scipy.fftpack import convolve

In [3]: convolve
Out[3]: <module 'scipy.fftpack.convolve' from '/home/br/repos/scipy/scipy/build-install/lib/python3.8/site-packages/scipy/fftpack/convolve.cpython-38-x86_64-linux-gnu.so'>

In [4]: from scipy.fftpack.convolve import convolve

In [5]: convolve
Out[5]: <function scipy.fftpack.convolve.convolve>
```

OK to keep them separately here @rgommers ?

Here are the doctest checking logs:

```
br@gonzales:~/repos/scipy/scipy/tools$ python doctest_public_modules.py -s fftpack -v

=============
scipy.fftpack
=============
scipy.fftpack.fft
scipy.fftpack.ifft
scipy.fftpack.fftn
scipy.fftpack.ifftn
scipy.fftpack.rfft
scipy.fftpack.irfft
scipy.fftpack.fft2
scipy.fftpack.ifft2
scipy.fftpack.diff
scipy.fftpack.tilbert
scipy.fftpack.itilbert
scipy.fftpack.hilbert
scipy.fftpack.ihilbert
scipy.fftpack.sc_diff
scipy.fftpack.cs_diff
scipy.fftpack.cc_diff
scipy.fftpack.ss_diff
scipy.fftpack.shift
scipy.fftpack.fftfreq
scipy.fftpack.rfftfreq
scipy.fftpack.fftshift
scipy.fftpack.ifftshift
scipy.fftpack.next_fast_len
scipy.fftpack.dct
scipy.fftpack.idct
scipy.fftpack.dst
scipy.fftpack.idst
scipy.fftpack.dctn
scipy.fftpack.idctn
scipy.fftpack.dstn
scipy.fftpack.idstn
TestResults(failed=0, attempted=82)

>>>> OK: doctests PASSED
br@gonzales:~/repos/scipy/scipy/tools$ python doctest_public_modules.py -s fftpack.convolve -v


======================
scipy.fftpack.convolve
======================
scipy.fftpack.convolve.convolve
scipy.fftpack.convolve.convolve_z
scipy.fftpack.convolve.init_convolution_kernel
TestResults(failed=0, attempted=0)
```

ev-br commented:
OK, the current status is 'getting closer' :-). I've rebased on main to pick up recent fixes by Warren and Pamphile (thanks!), and fixed the bulk of simple stuff. Some of those included judgement calls, so please review https://github.com/scipy/scipy/pull/16391/commits/390d202b3ca9483163e91c26693ed7418a3f9940 and below (commits are done for a commit-by-commit review). 

The CI log is at https://github.com/scipy/scipy/runs/7222341253?check_suite_focus=true
Several things in need of attention of module regulars:
- io.netcdf : there's something going on with the memmapped `netcdf_file` -- @rgommers IIRC you looked at these a while ago?
- differential_evolution : the `(nct)` part of  `differential_evolution(rosen, bounds, constraints=(nlc)` looks suspiciously like a missing comma, but adding it does not fix the error. @andyfaff could you weigh in? 
- optimize.nonlin stuff is tracked in gh-16532
- stats.anderson_ksamp warning is strange, could you give me a hand @mdhaber ?
- optimize tutorial / linprog part. There is a talk about `method='revised simplex'`, is the whole paragraph still relevant? 

And the elephant in the room is that the dev.py plumbing, which I copied from `refguide-check`, still reports success where it should not have. This is gh-16476, and I'll ask @rgommers . I tried to poke around the new CLI stuff but did not have much success.

WarrenWeckesser commented:
I'll submit a pull request soon to fix the `differential_evolution` warning.

rgommers commented:
> io.netcdf : there's something going on with the memmapped `netcdf_file` -- @rgommers IIRC you looked at these a while ago?

I don't see anything in the CI log nor in the PR diff, so just to confirm: this is gone, right?

> And the elephant in the room is that the dev.py plumbing, which I copied from `refguide-check`, still reports success where it should not have. This is [gh-16476](https://github.com/scipy/scipy/issues/16476), and I'll ask @rgommers . I tried to poke around the new CLI stuff but did not have much success.

I'll have a quick look, but if it's nontrivial it'll unfortunately have to wait till after the last build-related 1.9.0 release blocker is resolved.

ev-br commented:
(io.netcdf_file) 
> I don't see anything in the CI log nor in the PR diff, so just to confirm: this is gone, right?

I ended up doing 
https://github.com/scipy/scipy/pull/16391/commits/c4a3ad2a54076b77ecc1fbd41bd154df1fac83b6 , but I'm not sure it's actually the right thing to recommend to users. 
An alternative is of course to filter out the warning.

rgommers commented:
> I ended up doing
> [c4a3ad2](https://github.com/scipy/scipy/commit/c4a3ad2a54076b77ecc1fbd41bd154df1fac83b6) , but I'm not sure it's actually the right thing to recommend to users.

That seems right to me, and it's what the tests do as well.

ev-br commented:
Note to self: tutorial files are not only in `doc/source/tutorial`, also `doc/source/tutorial/stats` etc, so need to recurse into subdirs.

lucascolley commented:
@ev-br is this superseded by gh-19242 or is that only doing a subset of what's here?

ev-br commented:
@lucascolley mostly superseded, yes. I plan to take a closer look at gh-19242 in the next couple of weeks. Meanwhile, feel free to close this one if it gets in the way.

lucascolley commented:
superseded by gh-20127
